### PR TITLE
Change Halstead implementation

### PR DESCRIPTION
**Pull request by Luni-4**
In this way `CodeMetrics` is independent from lifetimes.

Thanks in advance for your review! :)

### CODE REVIEW

**Comment by calixteman** [comment]
Could you make a PR with the tests only and then we can see if the refactoring doesn't introduce any regression ?

**Comment by Luni-4** [comment]
> Could you make a PR with the tests only and then we can see if the refactoring doesn't introduce any regression ?

Yep, good idea, thanks! :)

File: src/metrics/halstead.rs [hunk 1]
```diff
@@ -76,39 +113,37 @@ impl<'a> fmt::Display for Stats<'a> {
...
+impl Stats {
     /// Merges a second `Halstead` metric suite into the first one
-    pub fn merge(&mut self, other: &Stats<'a>) {
-        for (k, v) in other.operators.iter() {
-            *self.operators.entry(*k).or_insert(0) += v;
-        }
-        for (k, v) in other.operands.iter() {
-            *self.operands.entry(*k).or_insert(0) += v;
-        }
+    pub fn merge(&mut self, other: &Stats) {
```
Comment (calixteman) [inline src/metrics/halstead.rs:118]: we shouldn't have this function.
You can't sum the u_***: the maps need to be merged and then the new values are the length of the new maps.

File: src/metrics/halstead.rs [hunk 2]
```diff
@@ -194,53 +228,104 @@ fn get_id<'a>(node: &Node<'a>, code: &'a [u8]) -> &'a [u8] {
...
+    code: &'a [u8],
+    stats: &mut Stats,
+    halstead_maps: &mut HalsteadMaps<'a>,
+) {
     match T::get_op_type(&node) {
-        HalsteadType::Operator => *stats.operators.entry(node.kind_id()).or_insert(0) += 1,
-        HalsteadType::Operand => *stats.operands.entry(get_id(node, code)).or_insert(0) += 1,
+        HalsteadType::Operator => {
+            *halstead_maps.operators.entry(node.kind_id()).or_insert(0) += 1;
+            stats.u_operators = halstead_maps.operators.len() as u64;
```
Comment (calixteman) [inline src/metrics/halstead.rs:240]: I think it's useless to update this value each time we insert a new operators.
Maybe we can add a `finalize` function called when we just finish parsing to set u_****.

File: src/metrics/halstead.rs [hunk 3]
```diff
@@ -194,53 +228,104 @@ fn get_id<'a>(node: &Node<'a>, code: &'a [u8]) -> &'a [u8] {
...
+            stats.operators += 1;
+        }
+        HalsteadType::Operand => {
+            *halstead_maps
+                .operands
+                .entry(get_id(node, code))
+                .or_insert(0) += 1;
+            stats.u_operands = halstead_maps.operands.len() as u64;
```
Comment (calixteman) [inline src/metrics/halstead.rs:248]: see above

**Comment by calixteman** [comment]
Please rebase.

**Comment by Luni-4** [comment]
The new `finalize` approach breaks the tests, it seems the `Halstead` is no more computed in the right way.

**Comment by Luni-4** [comment]
Now it works, I forgot to update the `Mi` implementation and update stacks :P

**Comment by Luni-4** [comment]
@calixteman 

Can we merge this PR, please?

**Pull request merged**